pkg/proctree/tree.go

```diff
@@ -0,0 +1,297 @@
+func (tree *ProcessTree) hasThread(hostThreadID int) bool {
+	_, exist := tree.threads.Get(hostThreadID)
+	return exist
+}
+
+// removeProcess remove the process node from the process tree
+// This does not remove references to it from other nodes and vice versa, nor allow it to be
+// garbage collected.
+func (tree *ProcessTree) removeThread(tnode *threadNode) error {
+	tree.threads.Delete(tnode.getId())
+	return nil
+}
+
+// cleanThread remove the thread from the GC LRU, triggering cleaning eviction function
+func (tree *ProcessTree) cleanThread(tnode *threadNode) error {
+	_ = tree.threadsGC.Remove(tnode.GetUniqueId())
+	return nil
+}
+
+// deleteThread remove the thread node from the process tree and delete all references to it,
+// so it could be garbage collected
+func (tree *ProcessTree) deleteThread(tnode *threadNode) error {
+	err := tree.removeThread(tnode)
+	if err != nil {
+		logger.Errorw("delete thread from tree", "err", err.Error())
+	}
+	tnode.DisconnectNode()
+	return nil
+}
+
+// addGeneralEventProcess generate a new process with information that could be received from any
+// event from the process
+func (tree *ProcessTree) addGeneralEventProcess(event *trace.Event) (*processNode, error) {
+	process, err := tree.newProcessNode(event.HostProcessID)
+	if err != nil {
+		return nil, err
+	}
+	process.setGenInfoFromEventProtected(event)
+	return process, nil
+}
+
+// addGeneralEventThread generate a new thread with information that could be received from any
+// event from the thread
+func (tree *ProcessTree) addGeneralEventThread(event *trace.Event) (*threadNode, error) {
+	p, err := tree.getProcess(event.HostProcessID)
+	if err != nil {
+		p, err = tree.addGeneralEventProcess(event)
+		if err != nil {
+			return nil, err
+		}
+	}
+	p.mutex.Lock()
+	thread, err := tree.newProcessThreadNode(p, event.HostThreadID)
+	exitTime := p.getExitTime()
+	p.mutex.Unlock()
+	if err != nil {
+		return nil, err
+	}
+
+	thread.setGenInfoFromEventProtected(event, exitTime)
+	return thread, nil
+}
+
+// generateParentProcess add a parent process to given process from tree if existing or creates
+// new node with the best effort info
+func (tree *ProcessTree) generateParentProcess(parentHostId int, parentNsId int, process *processNode) (*processNode, error) {
+	if parentNsId != 0 &&
+		process.getId() != parentHostId { // Prevent looped references
+		parentProcess, err := tree.getProcess(parentHostId)
+		if err != nil {
+			// TODO: Fix the race condition here between checking if exist and setting new one
+			parentProcess, err = tree.newProcessNode(parentHostId)
+			if err != nil {
+				return nil, err
+			}
+			parentProcess.setNsId(parentNsId)
+			if err != nil {
+				return nil, err
+			}
+		}
+		process.mutex.Lock()
+		process.connectParent(parentProcess)
+		process.mutex.Unlock()
+		parentProcess.mutex.Lock()
+		parentProcess.connectChild(process)
+		parentProcess.mutex.Unlock()
```

Comment:
Isn't this a clear case of a transaction ?

 Where parent and child both have dependency on each other ? I would assume intrinsically, in a first read, you would have to:

1. lock parent
2. lock process
3. connect process to parent
4. conenct parent to process
5. unlock process
6. unlock parent

this way if other childs are to be added to the parent process they would only be able to once parent is "consitent" (thats what defines a transaction consistency).

Reply:
So I wanted to avoid locking both locks at once, to avoid deadlocks in the long term.
However, you are right here that this is crucial for them to be consistent with each other.
I will change it.